File: leetcode/101-200/102-binary-tree-level-order-traversal/bench/level_order.c
// Benchmark workload for LeetCode #102 — level order, C mirror (malloc *Node).
// Build 8 BSTs once, then K reps of DFS-with-depth level_order on a data-dependent-
// selected tree, folding level count + each level's size + every value into a rolling
// hash. Each rep allocates a fresh result (an array of dynamically-grown int arrays,
// one per depth) and frees it — the allocation-bound churn the benchmark measures.
#include <stdio.h>
#include <stdlib.h>
#define MOD 1000000007LL

typedef struct Node { long long val; struct Node *left, *right; } Node;

static Node* insert(Node* root, long long v) {
    if (!root) { Node* n = malloc(sizeof(Node)); n->val = v; n->left = n->right = NULL; return n; }
    if (v < root->val) root->left = insert(root->left, v);
    else root->right = insert(root->right, v);
    return root;
}

// A growable row of values.
typedef struct { long long* data; long long len, cap; } Row;
// A growable result: rows indexed by depth.
typedef struct { Row* rows; long long len, cap; } Result;

static void row_push(Row* r, long long v) {
    if (r->len == r->cap) { r->cap = r->cap ? r->cap * 2 : 4; r->data = realloc(r->data, r->cap * sizeof(long long)); }
    r->data[r->len++] = v;
}
static void result_push_row(Result* res) {
    if (res->len == res->cap) { res->cap = res->cap ? res->cap * 2 : 4; res->rows = realloc(res->rows, res->cap * sizeof(Row)); }
    res->rows[res->len].data = NULL; res->rows[res->len].len = 0; res->rows[res->len].cap = 0;
    res->len++;
}

static void dfs(Node* node, long long depth, Result* res) {
    if (!node) return;
    if (depth == res->len) result_push_row(res);
    row_push(&res->rows[depth], node->val);
    dfs(node->left, depth + 1, res);
    dfs(node->right, depth + 1, res);
}

int main(void) {
    long long base[] = {16, 8, 24, 4, 12, 20, 28, 2, 6, 10, 14, 18, 22, 26, 30};
    long long bn = 15;
    Node* pool[8];
    for (long long t = 0; t < 8; t++) {
        Node* root = NULL;
        for (long long k = 0; k < bn; k++) root = insert(root, base[(k + t) % bn]);
        pool[t] = root;
    }
    long long acc = 1;
    for (long long rep = 0; rep < 1500000; rep++) {
        long long idx = acc % 8;
        Result res = { NULL, 0, 0 };
        dfs(pool[idx], 0, &res);
        acc = (acc * 131 + res.len) % MOD;
        for (long long li = 0; li < res.len; li++) {
            Row* lvl = &res.rows[li];
            acc = (acc * 131 + lvl->len) % MOD;
            for (long long vi = 0; vi < lvl->len; vi++)
                acc = (acc * 131 + lvl->data[vi]) % MOD;
            free(lvl->data);
        }
        free(res.rows);
    }
    printf("%lld\n", acc);
    return 0;
}

File: leetcode/101-200/102-binary-tree-level-order-traversal/bench/level_order.rs
// Benchmark workload for LeetCode #102 — level order, Rust mirror (Box + &-borrow).
// The tree is never mutated or shared during traversal, so a single-owner Box walked
// by reference is the honest Rust node. Build 8 BSTs once, then K reps of DFS-with-depth
// level_order on a data-dependent-selected tree, folding the whole Vec<Vec<i64>> result
// into a rolling hash. Each rep allocates a fresh Vec<Vec<i64>> and drops it.
const MOD: i64 = 1000000007;

struct Node { val: i64, left: Option<Box<Node>>, right: Option<Box<Node>> }

fn insert(root: Option<Box<Node>>, v: i64) -> Option<Box<Node>> {
    match root {
        None => Some(Box::new(Node { val: v, left: None, right: None })),
        Some(mut n) => {
            if v < n.val { n.left = insert(n.left.take(), v); }
            else { n.right = insert(n.right.take(), v); }
            Some(n)
        }
    }
}

fn dfs(node: &Option<Box<Node>>, depth: usize, result: &mut Vec<Vec<i64>>) {
    if let Some(n) = node {
        if depth == result.len() { result.push(Vec::new()); }
        result[depth].push(n.val);
        dfs(&n.left, depth + 1, result);
        dfs(&n.right, depth + 1, result);
    }
}

fn level_order(root: &Option<Box<Node>>) -> Vec<Vec<i64>> {
    let mut result: Vec<Vec<i64>> = Vec::new();
    dfs(root, 0, &mut result);
    result
}

fn main() {
    let base: [i64; 15] = [16, 8, 24, 4, 12, 20, 28, 2, 6, 10, 14, 18, 22, 26, 30];
    let bn = base.len();
    let mut pool: Vec<Option<Box<Node>>> = Vec::new();
    for t in 0..8 {
        let mut root: Option<Box<Node>> = None;
        for k in 0..bn { root = insert(root, base[(k + t) % bn]); }
        pool.push(root);
    }
    let mut acc: i64 = 1;
    for _ in 0..1500000i64 {
        let idx = (acc % 8) as usize;
        let levels = level_order(&pool[idx]);
        acc = (acc * 131 + levels.len() as i64) % MOD;
        for lvl in &levels {
            acc = (acc * 131 + lvl.len() as i64) % MOD;
            for &v in lvl {
                acc = (acc * 131 + v) % MOD;
            }
        }
    }
    println!("{}", acc);
}
